### PR TITLE
Move some of the grain instance initialization logic outside lock.

### PR DESCRIPTION
**Pull request by jthelin**
- Can move some of the grain instance initialization logic outside lock.

### CODE REVIEW

File: src/OrleansRuntime/Catalog/Catalog.cs [hunk 1]
```diff
@@ -590,8 +602,6 @@ private void CreateGrainInstance(string grainTypeName, ActivationData data, stri
 
             activations.IncrementGrainCounter(grainClassName);
 
-            data.GrainInstance.Data = data;
```
Comment (gabikliot) [inline src/OrleansRuntime/Catalog/Catalog.cs]: where did that line go?

**Comment by gabikliot** [comment]
Why is it important to take this code outside the lock? This lock is on this specific activation, and this activation anyway cannot do anything before it was initialized, so I don't see the advantage of reducing this lock scope. On the other hand, potentially, in theory, it can have a down side of now maybe not protecting some critical operation that should have been protected.

**Comment by gabikliot** [comment]
Looked at this PR again, more carefully. The lock inside `CreateGrainInstance` is actually **completely redundant**, since the call to `CreateGrainInstance` is already [inside a lock on this `ActivationData`](https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/Catalog/Catalog.cs#L406). So you can safely remove the inner lock, while leaving all the code in place, with no change. Or just leave the lock there, since it has zero impact on anything - if thread already owns a lock, re-taking is has basically zero overhead.

File: src/OrleansRuntime/Catalog/Catalog.cs (hunk 1 above)
Comment (jthelin) [inline src/OrleansRuntime/Catalog/Catalog.cs]: It is at line 575 now - trying to collect the two runtime injections into the grain instance into one place.
https://github.com/jthelin/orleans/commit/db5c3ead5b6ed342dff058f6f71cb1f3a2a844c2#diff-bcfe9efaf8ce10fd652fca0dd5a4d116R575

`grain` and `data.GrainInstance` are the same thing, but previous code doesn't make that very clear.

**Pull request merged**